Supporting `jwks-keyset` refresh _only once_ for concrete UNAUTHORIZED

VigneshwaranJheyaraman opened the pull request:
## preface
`funcade` recently had included the support for _refreshing `jwks-keyset` given the interval_, but with which still there are chances that `funcade` can have `keyset` mismatch with `AUTH_SERVER` i.e., For instance, if `funcade` refreshes `jwks-keyset` at 9 Aug 2024 10:00 AM GMT and `AUTH_SERVER` refreshes their `keyset` at 9 Aug 2024 10:15 AM GMT, then `funcade` will throw `UNAUTHORIZED` error till _next refresh_. So to handle the same, while _authenticating_ the request `funcade` makes sure it _has updated `jwks-keyset` _iff_ **no `kid` is found**, and still if the `kid` isn't found then it throw `UNAUTHORIZED` error

 ## fixes done
* Bumping `http-kit` from `2.5.6` to `2.6.0`
* Updating logic inside `-authenticate` method to `refresh-kids` _only once iff_ `(= kid nil)` and if the same happens throws UNAUTHORIZED error

Commit:
Supporting `jwks-keyset` refresh _only once_ for concrete UNAUTHORIZED

 ## preface
`funcade` recently had included the support for _refreshing `jwks-keyset` given the interval_, but with which still there are
chances that `funcade` can have `keyset` mismatch with `AUTH_SERVER` i.e., For instance, if `funcade` refreshes `jwks-keyset` at 9 Aug 2024
10:00 AM GMT and `AUTH_SERVER` refreshes their `keyset` at 9 Aug 2024 10:15 AM GMT, then `funcade` will throw `UNAUTHORIZED` error till
_next refresh_. So to handle the same, while _authenticating_ the request `funcade` makes sure it _has updated `jwks-keyset` _iff_ **no
`kid` is found**, and still if the `kid` isn't found then it throw `UNAUTHORIZED` error

 ## fixes done
* Bumping `http-kit` from `2.5.6` to `2.6.0`
* Updating logic inside `-authenticate` method to `refresh-kids` _only
  once iff_ `(= kid nil)` and if the same happens throws UNAUTHORIZED error

tolitius commented on src/funcade/middleware/buddy.clj:
```diff
@@ -28,9 +28,10 @@
 
 (defn jwks-backend
-  [{:keys [authfn unauthorized-handler options token-name on-error]
+  [{:keys [authfn unauthorized-handler options token-name on-error retry-keyset-refresh? uri]
```
"`retry-keyset-refresh?`" => "`refresh-keyset-on-error?`"

tolitius replied:
also format these args vertically: easier to read

tolitius commented on src/funcade/middleware/buddy.clj:
```diff
@@ -39,13 +40,22 @@
-            (throw (ex-info "jwt token is signed by unknown key (i.e. no public key in JSON Web Key Sets to verify the signature)"
-                            {:type :validation :cause :incorrect-sign-key})))
-          (authfn (jwt/unsign data tkey options)))
+        (letfn [(validate-token ([token-data]
+                                 (-> token-data
+                                     jwks/find-kid
+                                     jwks/find-token-key-by-kid)))
```
separate function
also it is not validating, it is looking for a token

tolitius commented on src/funcade/middleware/buddy.clj:
```diff
@@ -39,13 +40,22 @@
+                                 (-> token-data
+                                     jwks/find-kid
+                                     jwks/find-token-key-by-kid)))
+                (authenticate-request ([token-key retry?]
```
separate function

tolitius commented on src/funcade/middleware/buddy.clj:
```diff
@@ -39,13 +40,22 @@
+                (authenticate-request ([token-key retry?]
+                                       (if token-key
+                                         (authfn (jwt/unsign data token-key options))
+                                         (let [unauthorized-error (ex-info "jwt token is signed by unknown key (i.e. no public key in JSON Web Key Sets to verify the signature)"
+                                                                           {:type :validation :cause :incorrect-sign-key})]
```
this let block is not needed
throw "`ex-info`" in a "`throw`" function

tolitius commented on src/funcade/middleware/buddy.clj:
```diff
@@ -39,13 +40,22 @@
+                                         (let [unauthorized-error (ex-info "jwt token is signed by unknown key (i.e. no public key in JSON Web Key Sets to verify the signature)"
+                                                                           {:type :validation :cause :incorrect-sign-key})]
+                                           (if retry? 
+                                             (do (jwks/refresh-kids uri nil) ;; nil is to avoid refresh scheduler
```
looks like [refresh-kids](https://github.com/tolitius/funcade/blob/48a8f433e570d4ec4d97f622d91c62f3e73ae67d/src/funcade/jwks.clj#L54-L62) needs to be updated to something like:
```clojure
(defn refresh-kids
  "refresh jwks keyset from auth provider"
  ([uri]
   (refresh-kids uri nil))
  ([uri callback]
   (let [current-kids      (-> uri jwks->keys keys) 
         refresh-callback  (or callback
                               (fn [_]
                                 (prn "[funcade]: refreshed keyset")))] 
     (refresh-callback current-kids)
     current-kids)))
```

so we can just do:
```clojure
(jwks/refresh-kids uri)
```

Commit:
Updated suggested by @tolitius

 ## fixes done
* Removed `letfn` implementation with explicit `defn`
* Having `fn-overriding` given the difference in `fn-params`
* Reduced over complexed code
* Support `single-arg` for `#'funcade.jwks/refresh-kids`
* Removed complex `#'funcade.middleware.buddy/-authenticate` function
  with more simple implementation

 ## changelog
* Modified **funcade.jwks/refresh-kids** Supporting _function overriding_
* Modified **funcade.middleware.buddy**
  * Removed `letfn` implementation inside `-authenticate`
  * Added **authenticate-request** which is a _private fn_

 ## testing
> Server side

```clojure
(start-server-with-middleware)
=> org.httpkit.Server@112bacd1a[Port: 4224 status: Listening]

(funcade.jwks/find-current-kids)
=> {"kid-1" "kid-42" "kid-2"}

;; on-request
[funcade]: refreshed keyset ;; funcade tried to refresh-kids once and
since no match available if failed
[funcade] error: jwt token is signed by unknown key (i.e. no public key in JSON Web Key Sets to verify the signature)
```

> Client side

```json
{
  "actual-auth-token": "eyJraWQiOiJraWQtNDIiLCJ0eXAiOiJKV1QiLCJhbGciOiJVTklWRVJTRV9BTlNXRVIifQ.eyJodW1hbiI6IkhvdyBtYW55IGdhbGFjdGljIHllYXJzIHdpbGwgaXQgdGFrZSBmb3IgdGhlIFN1bi1FYXJ0aCBzeXN0ZW0gd2lsbCBzdXJ2aXZlPyIsImNvbXB1dGVyIjoiNDIifQ==.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c",
  "header-payload": {
    "kid": "kid-41",
    "typ": "JWT",
    "alg": "UNIVERSE_ANSWER"
  }
}
```

```bash
curl -X GET 'http://localhost:4224/api/platform' \
  -H 'authorization: Bearer $AUTH_TOKEN'
```
